src/frontend/src/optimizer/plan_expr_visitor/expr_counter.rs

```diff
@@ -0,0 +1,80 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+
+use crate::expr::{ExprImpl, ExprType, ExprVisitor, FunctionCall};
+
+/// `ExprCounter` is used by `CseRewriter`.
+#[derive(Default)]
+pub struct CseExprCounter {
+    // Only count pure function call and not const.
+    pub counter: HashMap<FunctionCall, usize>,
+}
+
+impl ExprVisitor<()> for CseExprCounter {
+    fn merge(_: (), _: ()) {}
+
+    fn visit_expr(&mut self, expr: &ExprImpl) {
+        // Considering this sql, `In` expression needs to ensure its in-clauses to be const.
+        // If we extract it into a common sub-expression (finally be a `InputRef`) which will
+        // violate this assumption, so ban this case. SELECT x,
+        //        tand(x) IN ('-Infinity'::float8,-1,0,1,'Infinity'::float8) AS tand_exact,
+        //        cotd(x) IN ('-Infinity'::float8,-1,0,1,'Infinity'::float8) AS cotd_exact
+        // FROM (VALUES (0), (45), (90), (135), (180),(225), (270), (315), (360)) AS t(x);
+        if expr.is_const() {
```

Comment:
`is_const` will perform a traversal on the given expression. Not sure if it hurts performance. 🥵

Reply:
Maybe we can enable this rule only for streaming queries.